substrate/frame/contracts/src/transient_storage.rs

```diff
@@ -0,0 +1,600 @@
+/// to the journal index (checkpoint) of when that call was entered is discarded.
+/// On `rollback_transaction`, all entries are reverted up to the last checkpoint.
+pub struct TransientStorage<T: Config> {
+	// The storage and journal size is limited by the storage meter.
+	current: Storage,
+	journal: Journal,
+	// The size of the StorageMeter is limited by the stack depth.
+	meter: StorageMeter<T>,
+	// The size of the checkpoints is limited by the stack depth.
+	checkpoints: Checkpoints,
+}
+
+impl<T: Config> TransientStorage<T> {
+	pub fn new(total_limit: u32, transaction_limit: u32) -> Self {
+		TransientStorage {
+			current: Default::default(),
+			journal: Journal::new(),
+			checkpoints: Default::default(),
+			meter: StorageMeter::new(total_limit, transaction_limit),
+		}
+	}
+
+	/// Read the storage entry.
+	pub fn read(&self, account: &AccountIdOf<T>, key: &Key<T>) -> Option<Vec<u8>> {
+		self.current.read(&account.encode(), &key.hash())
+	}
+
+	/// Write a value to storage.
+	pub fn write(
+		&mut self,
+		account: &AccountIdOf<T>,
+		key: &Key<T>,
+		value: Option<Vec<u8>>,
+		take: bool,
+	) -> Result<WriteOutcome, DispatchError> {
+		let prev_value = self.read(account, key);
+		// Skip if the same value is being set.
+		if prev_value != value {
+			let key = key.hash();
+			let account = account.encode();
+
+			// Calculate the allocation size.
+			if let Some(value) = &value {
+				// Charge the keys, value and journal entry.
+				// If a new value is written, a new journal entry is created. The previous value is
+				// moved to the journal along with its keys, and the new value is written to
+				// storage.
+				let keys_len = account.len().saturating_add(key.len());
+				let mut amount = value
+					.len()
+					.saturating_add(keys_len)
+					.saturating_add(mem::size_of::<JournalEntry>());
+				if prev_value.is_none() {
+					// Charge a new storage entry.
+					// If there was no previous value, a new entry is added to storage (BTreeMap)
+					// containing a Vec for the key and a Vec for the value. The value was already
+					// included in the amount.
+					amount = amount
+						.saturating_add(keys_len)
+						.saturating_add(mem::size_of::<Vec<u8>>().saturating_mul(2));
+				}
+				self.meter.charge(amount as _)?;
+			}
+			self.current.write(&account, &key, value);
+			// Update the journal.
+			self.journal.push(JournalEntry::new(&account, &key, prev_value.clone()));
+		}
+
+		Ok(match (take, prev_value) {
+			(_, None) => WriteOutcome::New,
+			(false, Some(prev_value)) => WriteOutcome::Overwritten(prev_value.len() as _),
+			(true, Some(prev_value)) => WriteOutcome::Taken(prev_value),
+		})
+	}
+
+	/// Remove a contract, clearing all its storage entries.
+	pub fn remove(&mut self, account: &AccountIdOf<T>) {
+		let account = account.encode();
+		// Remove all account entries.
+		let keys = self.current.keys(&account).collect::<Vec<_>>();
```

Comment:
we have to copy all of the keys for the contract. In a corner case, this can take up 70% of MaxTransientStorageLen. We have a margin of 2x, so it should still be safe.
There is also an option not to remove the storage on contract termination, as it will be deleted at the end of the transaction anyway